io-sim-classes/src/Control/Monad/Class/MonadFork.hs

```diff
@@ -8,16 +8,25 @@ import           Control.Exception
 import           Control.Monad (void)
 import           Control.Monad.Except
 import           Control.Monad.Reader
-import           System.IO (hPutStrLn, stderr)
+import           System.IO (hFlush, hPutStrLn, stderr)
+import           System.IO.Unsafe (unsafePerformIO)
+
+forkPrintExceptionLock :: IO.MVar ()
+{-# NOINLINE forkPrintExceptionLock #-}
+forkPrintExceptionLock = unsafePerformIO $ IO.newMVar ()
 
 class Monad m => MonadFork m where
   fork    :: m () -> m ()
 
 instance MonadFork IO where
   fork a =
     let handleException :: Either SomeException () -> IO ()
-        handleException (Left e) =
-            hPutStrLn stderr $ "Uncaught exception in thread:" ++ displayException e
+        handleException (Left e) = do
+            tid <- IO.myThreadId
+            IO.withMVar forkPrintExceptionLock $ \() -> do
+              hPutStrLn stderr $ "Uncaught exception in thread " ++ show tid
+                              ++ ": " ++ displayException e
+              hFlush stderr
```

Comment:
This isn't a criticism of the change here, but of having a default exception handler in general. With or without this change, the user has no way to stop this print to `stderr`. If you want exception logging, log the exception yourself, it should not be the default built-in unalterable definition of a `fork`.

Reply:
The point is that if you forget to add an exception handler, the default is to do nothing with it, and you will not notice. I agree that what is here is crude, but I personally find it invalueable when unexpected exceptions are noticed. Obviously printing them is crude, if we had better logging that would be better.